sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -734,30 +735,50 @@ object DataSource extends Logging {
    * Checks and returns files in all the paths.
    */
   private[sql] def checkAndGlobPathIfNecessary(
-      paths: Seq[String],
+      pathStrings: Seq[String],
       hadoopConf: Configuration,
       checkEmptyGlobPath: Boolean,
       checkFilesExist: Boolean): Seq[Path] = {
-    val allGlobPath = paths.flatMap { path =>
-      val hdfsPath = new Path(path)
-      val fs = hdfsPath.getFileSystem(hadoopConf)
-      val qualified = hdfsPath.makeQualified(fs.getUri, fs.getWorkingDirectory)
-      val globPath = SparkHadoopUtil.get.globPathIfNecessary(fs, qualified)
-
-      if (checkEmptyGlobPath && globPath.isEmpty) {
-        throw new AnalysisException(s"Path does not exist: $qualified")
-      }
+    val qualifiedPaths = pathStrings.map { pathString =>
+      val path = new Path(pathString)
+      val fs = path.getFileSystem(hadoopConf)
+      path.makeQualified(fs.getUri, fs.getWorkingDirectory)
+    }
+
+    // Split the paths into glob and non glob paths, because we don't need to do an existence check
+    // for globbed paths.
+    val (globPaths, nonGlobPaths) = qualifiedPaths.partition(SparkHadoopUtil.get.isGlobPath)
+
+    val globbedPaths = Try({
+      ThreadUtils.parmap(globPaths, "globPath", 20) { globPath =>
+        val fs = globPath.getFileSystem(hadoopConf)
+        val globResult = SparkHadoopUtil.get.globPath(fs, globPath)
+
+        if (checkEmptyGlobPath && globResult.isEmpty) {
+          throw new AnalysisException(s"Path does not exist: $globPath")
+        }
+
+        globResult
+      }.flatten
+    }).recoverWith({
+      case e: SparkException => throw e.getCause
+    }).get
+
 
-      // Sufficient to check head of the globPath seq for non-glob scenario
-      // Don't need to check once again if files exist in streaming mode
-      if (checkFilesExist && !fs.exists(globPath.head)) {
-        throw new AnalysisException(s"Path does not exist: ${globPath.head}")
+    try {
+      ThreadUtils.parmap(nonGlobPaths, "checkPathsExist", 20) { path =>
+        val fs = path.getFileSystem(hadoopConf)
+        if (checkFilesExist && !fs.exists(path)) {
```

Comment:
If checkFilesExist is false, we do not need to launch this parallel job actually.

Reply:
Good point!

Reply:
https://github.com/apache/spark/pull/25899/commits/7c61252fec9b23bcec173afe37a8d6db533a98d5